src/hdmf_zarr/zarr_utils.py

```diff
@@ -0,0 +1,289 @@
+"""
+Utilities for the HDF5 I/O backend,
+e.g., for wrapping HDF5 datasets on read, wrapping arrays for configuring write, or
+writing the spec among others"""
+
+from collections import deque
+from abc import ABCMeta, abstractmethod
+from collections.abc import Iterable
+from copy import copy
+
+from hdmf.build import (Builder,
+                        GroupBuilder,
+                        DatasetBuilder,
+                        LinkBuilder,
+                        BuildManager,
+                        RegionBuilder,
+                        ReferenceBuilder,
+                        TypeMap)
+
+import json
+import numpy as np
+import warnings
+import os
+import logging
+
+#from Zarrpy import Group, Dataset, RegionReference, Reference, special_dtype
+#from Zarrpy import filters as Zarrpy_filter
+from zarr import Array as ZarrArray
+
+from hdmf.array import Array
+from hdmf.query import HDMFDataset, ReferenceResolver, ContainerResolver, BuilderResolver
+from hdmf.utils import docval, getargs, popargs, get_docval
+
+
+class ZarrDataset(HDMFDataset):
+    @docval({'name': 'dataset', 'type': (np.ndarray, ZarrArray, Array), 'doc': 'the Zarr file lazily evaluate'},
+            {'name': 'io', 'type': 'ZarrIO', 'doc': 'the IO object that was used to read the underlying dataset'})
+    def __init__(self, **kwargs):
+        self.__io = popargs('io', kwargs)
+        super().__init__(**kwargs)
+
+    @property
+    def io(self):
+        return self.__io
+
+    @property
+    def shape(self):
+        return self.dataset.shape
+
+
+class DatasetOfReferences(ZarrDataset, ReferenceResolver, metaclass=ABCMeta):
```

Comment:
Aside from inheriting from `ZarrDataset` I think the functionality in this class is the same as in `DatasetOfReferences` in  HDF5. We may want to think about how we can avoid this duplication.

Reply:
Agreed. I think for now this is fine because I don't want to slow down a release on this next week on waiting for a HDMF ticket to be merged to support this change. I will make an issue ticket on HDMF still with the changes needed and a ticket on Zarr to make sure we update this once that has been merged. Timeline for that should be before Boston still as it shouldn't take long at all, that being said I don't want it slowing down this release. This is the same for the other comments you mentioned about refactoring to the backend in HDMF. @oruebel

Reply:
Agreed, creating issues for this is fine.